src/components/databaseUi/EventStoreUI.tsx

```diff
@@ -3,41 +3,104 @@ import React, { useState } from 'react';
-        <div className={ToolbarStyle.inputFieldRow}>
-          <label>Value: </label>
-          <input
-            className={ToolbarStyle.inputField}
-            type="text"
-            onChange={(e) => setInput(e.target.value)}
-          />
-        </div>
-        <div style={{display: 'flex', flexDirection: 'row', justifyContent: 'flex-end', marginTop: '0.5vh'}}>
-          <input type="submit" value="Add" className={ToolbarStyle.inputFieldSubmit}/>
-        </div>
-      </form>
-    </div>);
+    const [editorStatus, setEditorStatus] = useState({data: "", open: false});
+    const [errorStatus, setErrorStatus] = useState({error: false, msg:""});
+    const [activeField, setActiveField] = useState("");
+
+    function handleEditorCancel() {
+      setEditorStatus({...editorStatus, open: false});
+    }
+
+    async function handleEditorSubmit() {
+      let data;
+      try {
+        data = JSON.parse(editorStatus.data);
+      } catch (e) {
+        // Cannot be parsed as JSON. Pass as string instead.
+        data = editorStatus.data
+      }
+      
+      // Can put any JSON-stringifyable data. 
+      try {
+        await store.add(data);
+        setEditorStatus({...editorStatus, open: false});
+      } catch (error) {
+        setEditorStatus({...editorStatus, open: true});
+        setErrorStatus({error: true, msg: error});
+      }
+    }
+
+    const addEventEditor = () => {
+      return () => (
+        <Popup open={editorStatus.open}
+                 closeOnDocumentClick={false}
+                 contentStyle={{backgroundColor:"#0A3464FF"}}>
+          <div>
+            <AceEditor
+              mode="json"
+              theme="solarized_dark"
+              fontSize="2vh"
+              onChange={(data) => {
+                setEditorStatus({data: data, open: true});
+                setErrorStatus({...errorStatus, error:false});
+              }}
+              name="aceEditor"
+              value= {editorStatus.data}
+              width="50vw"
+              height="40vh"
+              editorProps={{ $blockScrolling: true }}
+            />
+            { errorStatus.error
+              ?
+              <div className={PopupEditorStyles.errorDiv}>
+                <label>{errorStatus.msg}</label>
+              </div>
+              :
+              <div className={PopupEditorStyles.buttonBar}>
+                <button className={ToolbarStyle.editButton} onClick={handleEditorCancel}>Cancel</button>
+                <button className={ToolbarStyle.editButton} onClick={handleEditorSubmit}>Submit</button>
+              </div>
+            }
+          </div>
+        </Popup>
+      );
     };
 
-    return (<div>
-      <div className={ToolbarStyle.buttonBar}>
-        <button className={ToolbarStyle.addButton} onClick={() => toggleAddFields(!showAddFields)}>Add</button>
+    const opTypes = {
+      Add: addEventEditor()
+    };
+
+    let op = "Add";
```

Comment:
If this never changes (doesn't look like it will), set it to const or simply inline it